designs/disruption-controls-by-method.md

```diff
@@ -0,0 +1,471 @@
+
+**Core**
+- NodePoolDrifted
+- RequirementsDrifted
+
+**AWS** 
+- AMIDrift
+- SubnetDrift 
+- SecurityGroupDrift 
+- NodeClassDrift
+
+**Azure** 
+- K8sVersionDrift 
+- ImageVersionDrift
+
+As you can see there are quite a few cases for drift, and the type of action that is taken from these forms of drift are very different. This leads to Drift Reasons needing a place in the api.
+
+#### Q: If all the reasons mainly apply to drift, why add a Method paired with reason rather than only allowing DriftReason and specifying that alongside drift only?
+Currently the method interface has a Type() implying that other methods also will want to specify disruption method types at a finer granularity 
+
+```go
+type Method interface {
+	ShouldDisrupt(context.Context, *Candidate) bool
+	ComputeCommand(context.Context, map[string]int, ...*Candidate) (Command, error)
+	Type() string
+	ConsolidationType() string
+}
+```
+
+
+
+#### Q: Why have the distinction between method and reason? Why not just have everything be a reason?
+The distinction between method and reason is crucial for providing both a high-level and a granular control over disruptions. The method corresponds to the type of disruption action, such as "Drift" or "Consolidation", which is a broad category of disruption. Within each method, there can be multiple reasons that provide specific context for the disruption, such as "AMIDrift" in the case of AWS.
+
+By separating method and reason, Karpenter allows users to define budgets and policies at both levels. Users can set a general budget for all "Drift" disruptions, and then further refine the control by specifying different budgets or schedules for specific reasons like "AMIDrift". This two-tiered approach offers flexibility and precision, enabling users to manage disruptions more effectively according to their needs.
+
+Moreover, the distinction helps in maintaining clarity and organization within the API. It allows for a structured way to handle disruptions, where methods can be seen as categories, and reasons as subcategories. This hierarchy makes it easier for users to navigate and understand the disruption policies they have set up.
+
+This also allows karpenter to easily tell inside of the disruption controller which actions it needs to be looking for when checking Type(). Without this top method, it becomes much more challenging to match a Type() with a Reason. 
+
+Alternatively we could do `method:reason`, then omit `:reason` to specify only the method granularity on a budget. But with this input coming from users, we will see more complex validation 
+
+#### Q: Budgets currently work by tracking deletion in total. In this new system, karpenter has to be aware of each disruption method + reason, does adding two fields Method + Reason make it harder to track how much of budget we have used? Should we add DisruptionReason to the nodeclaim? 
```

Comment:
Still thinking about this, can probably be consolidated with the "how should karpenter handle deletion by reason"